cmd/dockerd/daemon.go

```diff
@@ -779,3 +781,18 @@ func configureDaemonLogs(conf *config.Config) error {
 	})
 	return nil
 }
+
+func configureProxyEnv(conf *config.Config) {
+	if p := conf.HTTPProxy; p != "" {
+		_ = os.Setenv("HTTP_PROXY", p)
+		_ = os.Setenv("http_proxy", p)
+	}
+	if p := conf.HTTPSProxy; p != "" {
+		_ = os.Setenv("HTTPS_PROXY", p)
+		_ = os.Setenv("https_proxy", p)
+	}
+	if p := conf.NoProxy; p != "" {
+		_ = os.Setenv("NO_PROXY", p)
+		_ = os.Setenv("no_proxy", p)
+	}
```

Comment:
I'm probably overthinking this, but we likely want the precedence of these to remain familiar to users, right?  CLI flag > environment variable > config file?  I'm not sure what the best way to accomplish that would be (nor if it's even worth doing so).

(If we were able to resolve that disparity somehow, we could probably remove the `getConfigOrEnv` function. :grimacing:)

Reply:
Hmm.. good point. 

I recall we had discussions around that when `daemon.json` was implemented. Afaik, we currently treat `--some-option` and `configfile.some-option` to be at the same level (which is why the daemon produces a "conflicting options" error if both are set). IIRC, the motivation at the time was that `--some-option` would be usually set in a config-file as well (e.g. a systemd unit), and therefore to be at the same level of priority.


I _thought_ we made them both `> env-var`, but looks like for, e.g.`DOCKER_DRIVER`, we make the env-var take precedence, i.e. the following will use `overlay2`, not `vfs`;

```bash
DOCKER_DRIVER=overlay2 dockerd --debug --storage-driver=vfs
```

I thought it would make sense to make the config (or --flag) > env-var to allow overriding any default proxy that's set in the environment, but perhaps that was the wrong idea 🤔

Reply:
^^ discussing the above in the maintainers meeting, and we'll keep the current implementation (allowboth `--flag` and `daemon.json` to override the "system" configuration), but if we find that a proxy env-var is present, we'll print a warning in the daemon logs.